sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SerializationConstructorsShouldBeSecured.cs

```diff
@@ -0,0 +1,139 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class SerializationConstructorsShouldBeSecured : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4212";
+        private const string MessageFormat = "Secure that serialization constructor.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly AttributeComparer attributeComparer = new AttributeComparer();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(FindPossibleViolations,
+                SyntaxKind.ConstructorDeclaration);
+        }
+
+        private void FindPossibleViolations(SyntaxNodeAnalysisContext c)
+        {
+            var constructorSyntax = (ConstructorDeclarationSyntax)c.Node;
+            var reportLocation = constructorSyntax?.Identifier.GetLocation();
+            if (reportLocation == null)
+            {
+                return;
+            }
+
+            var serializationConstructor = c.SemanticModel.GetDeclaredSymbol(constructorSyntax);
+            if (!serializationConstructor.IsSerializationConstructor())
+            {
+                return;
+            }
+
+            var classSymbol = serializationConstructor.ContainingType;
+            if (!classSymbol.Implements(KnownType.System_Runtime_Serialization_ISerializable))
```

Comment:
`ContainingType` could be null, is `Implements` checking for null?

Reply:
yes